pkg/gofr/datasource/mongo/mongo.go

```diff
@@ -77,27 +78,103 @@ func (c *Client) UseTracer(tracer any) {
 }
 
 // Connect establishes a connection to MongoDB and registers metrics using the provided configuration when the client was Created.
-func (c *Client) Connect() {
+func (c *Client) Connect(ctx context.Context) error {
 	c.logger.Logf("connecting to mongoDB at %v to database %v", c.config.URI, c.config.Database)
 
-	uri := c.config.URI
+	uri := c.getURI()
 
-	if uri == "" {
-		uri = fmt.Sprintf("mongodb://%s:%s@%s:%d/%s?authSource=admin",
-			c.config.User, c.config.Password, c.config.Host, c.config.Port, c.config.Database)
+	client, err := c.createClient(ctx, uri)
+	if err != nil {
+		return err
 	}
 
-	m, err := mongo.Connect(context.Background(), options.Client().ApplyURI(uri))
+	if err := c.pingDatabase(ctx, client); err != nil {
+		return err
+	}
+
+	c.setupMetrics()
+
+	c.Database = client.Database(c.config.Database)
+
+	return c.verifyDatabaseAccess(ctx)
+}
+
+func (c *Client) getURI() string {
+	if c.config.URI != "" {
+		return c.config.URI
+	}
+
+	return fmt.Sprintf("mongodb://%s:%s@%s:%d/%s?authSource=admin",
+		c.config.User, c.config.Password, c.config.Host, c.config.Port, c.config.Database)
+}
+
+func (c *Client) createClient(ctx context.Context, uri string) (*mongo.Client, error) {
+	clientOpts := options.Client().ApplyURI(uri)
+
+	client, err := mongo.Connect(ctx, clientOpts)
 	if err != nil {
-		c.logger.Errorf("error connecting to mongoDB, err:%v", err)
+		return nil, c.handleConnectionError(err)
+	}
+
+	return client, nil
+}
+
+func (c *Client) handleConnectionError(err error) error {
+	if c.isAuthenticationError(err) {
+		return fmt.Errorf("%w: %w", ErrAuthentication, err)
+	}
+
+	if c.isTimeoutError(err) {
+		return fmt.Errorf("%w: connection timeout", ErrGenericConnection)
+	}
+
+	return fmt.Errorf("%w: %w", ErrGenericConnection, err)
+}
+
+func (*Client) isTimeoutError(err error) bool {
+	return strings.Contains(err.Error(), "connection timeout") || mongo.IsTimeout(err)
+}
 
-		return
+func (*Client) isAuthenticationError(err error) bool {
+	return strings.Contains(err.Error(), "authentication failed") ||
+		strings.Contains(err.Error(), "AuthenticationFailed")
+}
```

Comment:
I have a big problem with this

Validating errors with what they contains is a bad idea

If you have no other way to do that, then you have to add a clear comment about it

Reply:
Take a look at the code

https://github.com/mongodb/mongo-go-driver/blob/b823409f1094fc54099bf64799a1e5a1d9e6fa67/x/mongo/driver/auth/auth.go#L148

An error could be caught with an errors.As, looking for auth.Error

https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.17.1/x/mongo/driver/auth#Error

Reply:
If we write new code which does not adhere to best practices, it should not be treated as outside the scope. Error validation should not be based on string matching.